### ISSUE TITLE
Expose setting custom cookie attributes for OAuth-related cookies

### ISSUE THREAD

**Issue by srikanthchelluri** [jupyterhub/jupyterhub#4905]
### Proposed change

We have a use case where we run JupyterHub inside an iframe within a particular page. When a user is not logged in, the user is redirected to sign in via our OAuth2 provider and is then redirected back - all within the same domain since the redirects for this initial login happen within the iframe

There are two response redirects and each attempt to set particular, relevant cookies via the `Set-Cookie` response headers:

- `/hub/oauth_login` sets `oauthenticator-state` [1]
- `/hub/oauth_callback` sets `jupyterhub-session-id` [2] and `jupyterhub-hub-login` [3]

However, these cookies are never registered by the browser - the error from browser development tools describes why:

```
The Set-Cookie header didn't specify a "SameSite" attribute and was defaulted to "SameSite=Lax" and was blocked because it came from a cross-site response which was not the response to a top-level navigation. The Set-Cookie had to have been set with "SameSite=None" to enable cross-site usage.
```

And so this results in the user getting a "400 Bad Request: OAuth state missing from cookies" on the `/hub/oauth_callback` request

Note that when the user is already logged in and there are no interactive redirects, the user access JupyterHub seamlessly

To resolve this, we set `SameSite=None` and `Secure` attributes on the cookie by defining custom handlers. Specifically:

- A custom handler extends `OAuthLoginHandler` and overrides method `set_state_cookie`
- A custom handler extends `OAuthCallbackHandler` and overrides methods `set_session_cookie` and `set_hub_cookie`

Each of these methods pass two additional kwargs to `_set_cookie` [4]: `secure=True`, `samesite='none'`. We're mainly okay doing this because we front our JupyterHub with a proxy that allows only our trusted domain to embed the iframe via a content security policy. But rather than overriding entire methods, it would be preferable to have such settings configurable

Fundamentally, the ask is to expose settings/properties that allow users to configure custom cookie attributes for each kind of cookie (or describe why this isn't reasonable and provide possible alternatives)

### Alternative options

As described above, the way we're doing it now is by overriding the methods entirely so we can set those custom attributes in the invocations to `_set_cookie`

We could also just re-write out iframe flow to use top-level redirects for authentication

### Who would use this feature?

This should help with any use case that requires setting custom attributes on cookies

### References

[1] https://github.com/jupyterhub/oauthenticator/blob/main/oauthenticator/oauth2.py
[2] https://jupyterhub.readthedocs.io/en/4.1.2/tutorial/getting-started/security-basics.html#jupyterhub-session-id
[3] https://jupyterhub.readthedocs.io/en/4.1.2/tutorial/getting-started/security-basics.html#jupyterhub-hub-login
[4] https://github.com/jupyterhub/jupyterhub/blob/main/jupyterhub/handlers/base.py

**Comment by minrk** [comment]
Have you tried:

```python
c.JupyterHub.tornado_settings = {
    "cookie_options": dict(samesite="None", secure=True),
}
``` 

?

**Comment by srikanthchelluri** [comment]
@minrk Yup, that works - thanks! Curious if there’s any value in exposing the ability to set attributes on specific cookies (e.g. only want attributes X, Y to apply to cookie Z) - but for now this is certainly simpler than overriding handlers ourselves

Thanks for the quick response!

**Comment by minrk** [comment]
I'm not aware of a case where folks need that, so I'll wait for it to be needed to implement it. So far, it's been specific cookies needing specific attributes, but all cookies having them is fine, too.
